colossalai/checkpoint_io/distributed_checkpoint_utils.py

```diff
@@ -0,0 +1,511 @@
+import json
+import logging
+import os
+from pathlib import Path
+from typing import Dict, Iterator, Optional, OrderedDict, Tuple
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+from torch.distributed.distributed_c10d import _get_default_group
+
+from colossalai.interface import ModelWrapper
+from colossalai.utils import get_non_persistent_buffers_set
+
+from .index_file import CheckpointIndexFile
+from .utils import (
+    StateDictSharder,
+    async_save_state_dict_shards,
+    create_pinned_state_dict,
+    get_model_base_filenames,
+    load_state_dict,
+    save_state_dict,
+    save_state_dict_shards,
+    search_tp_partition_dim,
+)
+
+try:
+    from torch.nn.modules.module import _EXTRA_STATE_KEY_SUFFIX
+except ImportError:
+    _EXTRA_STATE_KEY_SUFFIX = "_extra_state"
+
+MODEL_META_PREFIX = "pytorch_model-meta-dist-"
+MODEL_WEIGHT_PREFIX = "pytorch_model-dist-"
+SHARD_META_SUFFIX = ".index.json"
+
+
+def dist_model_state_dict(model: nn.Module, prefix: str = "", keep_vars: bool = False):
+    destination = dict()
+    # Save parameters.
+    for name, param in model.named_parameters():
+        if param is None:
+            continue
+        destination[prefix + name] = param
+    # Save buffers.
+    non_persist_buffers_set = get_non_persistent_buffers_set(model)
+    for name, buf in model.named_buffers():
+        if buf is not None and name not in non_persist_buffers_set:
+            buffer = buf if keep_vars else buf.detach()
+            destination[prefix + name] = buffer
+
+    # Save extra states.
+    extra_state_key = prefix + _EXTRA_STATE_KEY_SUFFIX
+    if (
+        getattr(model.__class__, "get_extra_state", torch.nn.Module.get_extra_state)
+        is not torch.nn.Module.get_extra_state
+    ):
+        extra_state = model.get_extra_state()
+        destination[extra_state_key] = extra_state
+    return destination
+
+
+def load_state_dict_into_dist_model(
```

Comment:
What is this function for? Is it for loading whole state dict? Default `model.load_state_dict()` has already implemented this feature.